src/zkevm_specs/evm/execution/sar.py

```diff
@@ -0,0 +1,185 @@
+from ...util import (
+    FQ,
+    MAX_U64,
+    N_BYTES_U64,
+    RLC,
+    int_is_neg,
+    word_to_lo_hi,
+)
+from ..instruction import Instruction, Transition
+from ..typing import Sequence
+
+
+def sar(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+
+    shift = instruction.stack_pop()
+    a = instruction.stack_pop()
+    b = instruction.stack_push()
+
+    (
+        a64s,
+        b64s,
+        a64s_lo,
+        a64s_hi,
+        shf_div64,
+        shf_mod64,
+        p_lo,
+        p_hi,
+        p_top,
+    ) = gen_witness(instruction, shift, a)
+    check_witness(
+        instruction,
+        shift,
+        a,
+        b,
+        a64s,
+        b64s,
+        a64s_lo,
+        a64s_hi,
+        shf_div64,
+        shf_mod64,
+        p_lo,
+        p_hi,
+        p_top,
+    )
+
+    instruction.step_state_transition_in_same_context(
+        opcode,
+        rw_counter=Transition.delta(3),
+        program_counter=Transition.delta(1),
+        stack_pointer=Transition.delta(1),
+    )
+
+
+def check_witness(
+    instruction: Instruction,
+    shift: RLC,
+    a: RLC,
+    b: RLC,
+    a64s: Sequence[FQ],
+    b64s: Sequence[FQ],
+    a64s_lo: Sequence[FQ],
+    a64s_hi: Sequence[FQ],
+    shf_div64: FQ,
+    shf_mod64: FQ,
+    p_lo: FQ,
+    p_hi: FQ,
+    p_top: FQ,
+):
+    is_neg, _ = instruction.compare(FQ(127), FQ(a.le_bytes[31]), 1)
+    shf_lo, shf_hi = instruction.word_to_lo_hi(shift)
+    shf_hi_is_zero = instruction.is_zero(shf_hi)
```

Comment:
this part is inaccurate. If the `shf[1..32]` is non-zero, which means that `shift >= 256`, we should consider the shift operation is overflowed. Right now, this only checks if the high-128bit is zero or not.

Reply:
It seems that the current implementation could handle this case (and `shf_hi_is_zero` is not equal to `shift < 256`). Since `shf_lo` could be greater than or equal to 4.